### PR TITLE
Nonce simplification

### PR DESCRIPTION
**Pull request by arik-so**

### CODE REVIEW

**Comment by Roasbeef** [comment]
Hey can y'all remove that first commit from the branch? It ends up breaking the static remote key semantics, which are important to retain IMO. I've commented on that PR separately. Once that's gone I think we can have this land, and I'll do a few other corrections. I think this also might be based off a dated version of my original branch?

File: bolt-simple-taproot.md
Comment (Roasbeef) [inline bolt-simple-taproot.md:395]: I still think having _some_ sort of qualifier here that adds more context to the nonces sent would be more useful. 

For instance, this is Bob (responder of the channel), sending his local nonce which is used to allow Alice to sign the first commitment transaction. We can add more context by calling this either the remote nonce, or Bob's verification nonce.

File: bolt-simple-taproot.md
Comment (Roasbeef) [inline bolt-simple-taproot.md]: I don't think we actually need to add such slow level requirements. Today it's assumed that anything sent in the funding messages needs to be held in RAM until the process is complete. This includes all the relevant keys and constants sent along.

File: bolt-simple-taproot.md
Comment (Roasbeef) [inline bolt-simple-taproot.md]: I think this should read something like "public nonce the responder wants the initiator to use when signing their local commitment".

File: bolt-simple-taproot.md
Comment (Roasbeef) [inline bolt-simple-taproot.md:417]: Similar comment here re context: w/o loading the entire protocol into ones head, it's unclear what `public_nonce` actually even means here.

Comment (Roasbeef) [reply bolt-simple-taproot.md:417]: Continuing w/ the above, Alice is sending either her local nonce or verification nonce here. 

I still think the verification+signing stuff works pretty well. With this new formulation in the PR: 
  * signatures include one's signing nonce
  * some messages also send over the peer's verification nonce

Comment (Roasbeef) [reply bolt-simple-taproot.md:417]: AFAICT, the main goal here is to send nonces only at the very last step. As a result, we end up sending 2 nonces here (Alice's signing nonce and her verification nonce). 

Is that _really_ better than having her send it at the very start in `open_channel`? Then at this point, all nonces are already known and only a signature needs to be sent. 

IMO, the requirements w.r.t holding this in memory until the flow complete are more or less identical.

File: bolt-simple-taproot.md
Comment (Roasbeef) [inline bolt-simple-taproot.md]: If this is the public nonce, then it doesn't matter much if it gets written to disk or not. If not then wording doesn't make it clear if it's the secret vs public nonce.

File: bolt-simple-taproot.md
Comment (Roasbeef) [reply bolt-simple-taproot.md:417]: Also re the partial signature thing: originally the idea was that if we have a field, then why not just use the whole thing. Hence the idea to attach something to the 32-byte partial signature to round things out.

If nonces are known ahead of time, then that can just have empty bytes at the end. 

The other week you mentioned preferring not to have "contextual parsing", but that's already the case given that you need to know what features and the types of the TLV fields to parse this new message (compared to knowing that the sig will just be 32 bytes).

File: bolt-simple-taproot.md
Comment (Roasbeef) [inline bolt-simple-taproot.md]: This sort of crams a receiver's requirement (use this nonce to sign their commitment) in the same section as the set of sender requirements.

File: bolt-simple-taproot.md
Comment (Roasbeef) [inline bolt-simple-taproot.md]: Same here re low level requirements.

File: bolt-simple-taproot.md
Comment (Roasbeef) [inline bolt-simple-taproot.md]: Same here re public vs secret.

File: bolt-simple-taproot.md
Comment (Roasbeef) [inline bolt-simple-taproot.md:452]: Same here re semantics of nonce being sent.

File: bolt-simple-taproot.md
Comment (Roasbeef) [inline bolt-simple-taproot.md:465]: Actually a receiver requirement? 

In any case, no further commitments are signed until funding locked is sent.

File: bolt-simple-taproot.md
Comment (Roasbeef) [inline bolt-simple-taproot.md:487]: Same here re what nonce is being sent. Assuming this is one's local/verification nonce tho

File: bolt-simple-taproot.md
Comment (Roasbeef) [inline bolt-simple-taproot.md]: I think this should be broken apart, since the requirements for `channel_reestablish` are distinct as they need to factor in the fact that for retransmissions, one needs to re-generate any signatures that need retransmission with the new nonces.

File: bolt-simple-taproot.md
Comment (Roasbeef) [inline bolt-simple-taproot.md:572]: The nice thing about sending it in the `shutdown` message is that you effectively create an entirely new musig2 session. Otherwise, you'll need to thread through that last living session into w/e handles the co-op close state machine. 

At this point we're entering a distinct phase of the protocol, so why not send distinct nonces as well? You also don't need to worry about weird edge cases like sending a commitment sig, then accidentally re-using it when you go to down (consider that they never got the sig -- or so they claim, then you decide not to send it and just co-op close).

File: bolt-simple-taproot.md
Comment (Roasbeef) [inline bolt-simple-taproot.md:532]: This requirement should be carried above re the old sig field in the funding signed/created messages.

File: bolt-simple-taproot.md
Comment (Roasbeef) [inline bolt-simple-taproot.md]: The funding pub key is already known at this point, no need to re-compute it.

File: bolt-simple-taproot.md
Comment (Roasbeef) [inline bolt-simple-taproot.md]: Potentially just goes away if we send nonces in `shutdown`.

File: bolt-simple-taproot.md
Comment (Roasbeef) [inline bolt-simple-taproot.md]: It's a public nonce, so no harm really, no need to prescribe something so low level.

File: bolt-simple-taproot.md
Comment (Roasbeef) [inline bolt-simple-taproot.md]: Same here: _it is known_.

File: bolt-simple-taproot.md
Comment (Roasbeef) [inline bolt-simple-taproot.md:572]: I think _this_ change to the commitment dance is much more compelling that the initial funding changes. 

Before it was:
  * both sides exchange signing+verification nonces at the start of the connection
  * when you sign, you send your next signing nonce (you'll use to sign state N+1)
  * when they revoke, they send their next verification nonce (you'll use to sign state N+1)

With this it's:
  * both sides exchange a verification/local nonce at the start 
  * you sign sign, you send your signing nonce (lets them verify it)
  * when they revoke, they send their next verification nonce

Comment (Roasbeef) [reply bolt-simple-taproot.md:572]: Typing it out then reading it back, doesn't seem as striking, since the revoke is more or less the same, and you're still sending a nonce w/ the sig, just that it's the next/current nonce instead.

File: bolt-simple-taproot.md
Comment (Roasbeef) [inline bolt-simple-taproot.md:612]: local/verification

File: bolt-simple-taproot.md
Comment (Roasbeef) [inline bolt-simple-taproot.md:679]: I don't think this section can just be removed all together. 

As is, these changes don't mention how to handle retransmission at at all. Remember that it's possible that a node sends a state during retransmission, but with a distinct set of HTLCs (one of them timed out or w/e). If that _same_ nonce is used to send this extra sig, then the other party can lie that they didn't get the sig and force you to re-use the nonce.

**Review by Roasbeef — CHANGES_REQUESTED** [review]

**Comment by Roasbeef** [comment]
So I did a proper pass over everything, and came away not really grasping why the JIT nonce is so great for the _sender_. They still need to write their local/verification nonce to disk in order to generate a signature, if the nonce isn't counter based. 

I also think we should retain the local/verification nonce terminology _somewhere_, since it makes what's being sent self describing. We can just say `signing_nonce` (defined at the top) and be done w/ it, instead of saying listing `musig2_nonce`, then: "the sender must send the nonce they used to sign the commitment to the remote party.

I still think we should just prescribe that a counter be used, since it means that no new cryptographic material needs to be written to disk for _this_ flavor of taproot channels. But ofc we can do this in another change, as ideally we add a distinct section re nonce assumptions and recommendations, etc.

File: bolt-simple-taproot.md
Comment (Roasbeef) [reply bolt-simple-taproot.md:417]: > Is that really better than having her send it at the very start in open_channel? Then at this point, all nonces are already known and only a signature needs to be sent.

Also today after `accept_channel` is sent, all the necessaray fundign params have been exchanged (channel config effectively static). With this, you continue to send funding params even up until the very last message in the funding flow.

File: bolt-simple-taproot.md
Comment (wpaulino) [reply bolt-simple-taproot.md:395]: Not opposed to doing so, but the idea was for it to be clear based on the TLV type/field alone, since local/verification nonces will always be sent on their own, and remote/signing nonces will always be sent along with the partial sig.

File: bolt-simple-taproot.md
Comment (wpaulino) [reply bolt-simple-taproot.md]: We could tack that on as an additional requirement perhaps? We should still make it clear that the nonce the sender is sending standalone is the their nonce when signing their commitments.

File: bolt-simple-taproot.md
Comment (wpaulino) [reply bolt-simple-taproot.md:417]: One of the main reasons behind our proposal is that (in our opinion) it is much easier to reason about which nonce needs to be sent when and how should it be used. As @arik-so and I tried to discuss the how the protocol changes worked, we realized the local/remote and verification/signing terminology work well for one side but break down for the other, so we avoided them completely.

> Also re the partial signature thing: originally the idea was that if we have a field, then why not just use the whole thing. Hence the idea to attach something to the 32-byte partial signature to round things out.
If nonces are known ahead of time, then that can just have empty bytes at the end.
The other week you mentioned preferring not to have "contextual parsing", but that's already the case given that you need to know what features and the types of the TLV fields to parse this new message (compared to knowing that the sig will just be 32 bytes).

I agree with you on the "contextual parsing" part, but the reason for the TLV is due to the simpler nonce interpretation. We were forced to use one since we're coupling the partial sig with a public nonce, so they wouldn't fit in the message's existing sig field. If we choose to forgo that, and commit everything upfront as previously done, then I'm not opposed to reusing the field for the partial sig and keeping the remaining 32 bytes empty.

> Also today after accept_channel is sent, all the necessaray fundign params have been exchanged (channel config effectively static). With this, you continue to send funding params even up until the very last message in the funding flow.

I don't see that as a big deal, I would rather keep the simpler interpretation of nonces over that.

File: bolt-simple-taproot.md
Comment (wpaulino) [reply bolt-simple-taproot.md]: It is a public nonce, we're just making it clear it should not be persisted to disk since the counterparty will propose a new nonce upon restarting/reconnecting. This is the case for all occurrences of this requirement.

File: bolt-simple-taproot.md
Comment (wpaulino) [reply bolt-simple-taproot.md]: We were trying to describe the purpose of the nonce from the sender's perspective. We can likely remove the second sentence and reword it to clarify such.

File: bolt-simple-taproot.md
Comment (wpaulino) [reply bolt-simple-taproot.md:465]: Not quite, the sender of `funding_signed` (channel acceptor/non-initiator) must send its partial signature as well for the initial commitment transaction. The channel initiator already sent theirs in `funding_created`.

File: bolt-simple-taproot.md
Comment (wpaulino) [reply bolt-simple-taproot.md]: I don't immediately see the difference, each side sends new nonces in `channel_reestablish` (these are local/verification since they're not sent along a partial sig) to their counterparty, allowing either side to retransmit any necessary updates along with their partial sig (which includes a new remote/signing nonce as well).

File: bolt-simple-taproot.md
Comment (wpaulino) [reply bolt-simple-taproot.md:572]: Similar to my reply on your `channel_reestablish` comment, the partial sigs sent to counterparties always use unique and randomly generated nonces, they should never be stored by the sender, so reuse should not happen.

File: bolt-simple-taproot.md
Comment (wpaulino) [reply bolt-simple-taproot.md:679]: The nonces sent along with the partial sigs are always unique and randomly generated, they should not be derived deterministically for that exact reason. The standalone nonces can be derived deterministically, but the partial sigs for those are never sent to the counterparty, so it's safe to reuse that specific nonce.

File: bolt-simple-taproot.md
Comment (arik-so) [reply bolt-simple-taproot.md]: Or, tacking on to @Roasbeef's comment, it could also stem from the name if it were something like `next_musig_nonce` or `next_verification_nonce`. Admittedly, I find "verification" to be a somewhat confusing term, so would personally prefer `next_local_nonce`, should we opt to go this route.

File: bolt-simple-taproot.md
Comment (arik-so) [reply bolt-simple-taproot.md]: Doesn't the protocol allow for a lengthy back-and-forth of `closing_signed` messages? How would a nonce in `shutdown` obviate the requirement for fresh nonces in these signatures?

File: bolt-simple-taproot.md
Comment (Roasbeef) [reply bolt-simple-taproot.md:395]: Gotcha, then perhaps we can define the ints used for the types above in the preliminary section, for that to flow elsewhere?

File: bolt-simple-taproot.md
Comment (Roasbeef) [reply bolt-simple-taproot.md]: > sender is sending standalone is the their nonce when signing their commitments.

a.k.a their signing nonce? ;)

File: bolt-simple-taproot.md
Comment (Roasbeef) [reply bolt-simple-taproot.md:417]: > I would rather keep the simpler interpretation of nonces over that

So I think this is the main thing I'm contesting here: is sending nonces _later_ in the process actually simpler? Or is it just shifting things around w/ minimal gain?

As an example, in the new flow, Alice doesn't send her signing+verification nonce until the 3rd message is exchanged. But there's no reason why she can't just send that in `open_channel` since there're no hard dependancies on her generating the nonce (the new musig 1.0 mixes in the private key _always_ she gets randomness augmentation by default). 

I think the argument is stronger w/ the state machine changes though, since Bob doesn't need to cache Alice's "next" signing nonce, as it'll just be delivered to him w/ the next sig.

File: bolt-simple-taproot.md
Comment (Roasbeef) [reply bolt-simple-taproot.md]: The difference is that for channel_read/funding_locked there's no retransmission, as no state updates have happened yet. For chan_reest, both sides _must_ regenerate their signatures using the newly exchanged nonces (gen new signing nonce, use received verification nonce of the remote party). 

It's worth pointing out this difference in chan_reest, as today, to retransmit a sig you just need to read it from disk and send it over. With these changes, you _must_ regenerate it given the new nonce, since we don't want to force either side to write to disk their unused nonces (otherwise I'd need to remember which nonce I sent you before the D/C to be able to make the final sig).

File: bolt-simple-taproot.md
Comment (Roasbeef) [reply bolt-simple-taproot.md:572]: > The nice thing about sending it in the shutdown message is that you effectively create an entirely new musig2 session. Otherwise, you'll need to thread through that last living session into w/e handles the co-op close state machine.

What do you think about this point? If we send in `shutdown`, then w/e handles that state machine can generate its own nonces. Otherwise the nonces need to be passed down thru w/e layers/connections, which may potentially require stronger synchronization.

File: bolt-simple-taproot.md
Comment (Roasbeef) [reply bolt-simple-taproot.md]: Yeah that's the case today, but the plan is to actually make it turn based: https://github.com/Roasbeef/lightning-rfc/pull/2 + https://github.com/lightning/bolts/pull/1016

So each side takes turns sending signatures, which actually makes some of the edge cases around `fee_range` interpretation easier to handle (new message to be able to reject/abort negotiation instead of silent/undefined failure). 

Sending it doesn't obviate the need to generate those signatures. But it _does_ allow the shutdown and normal channel operation to not have to share any nonce context ("_previously_ received in the latest").

File: bolt-simple-taproot.md
Comment (Roasbeef) [reply bolt-simple-taproot.md:572]: I think one nice part about sending the nonce along w/ the sig at the final step is: the message itself can be passed as aux input to the nonce generation function, binding the nonce to the commitment transaction itself (see prior point about being able to retransmit w/ diff HTLCs).

File: bolt-simple-taproot.md
Comment (arik-so) [reply bolt-simple-taproot.md:395]: You mean the TLV ID ints? Sure, we can, but for implementers that might also involve a lot of scrolling back and forth. I think I'm personally in favor of duplication, but I don't have a strong opinion.

File: bolt-simple-taproot.md
Comment (arik-so) [reply bolt-simple-taproot.md]: Wilmer has much stronger intuition than I do for the nonce naming. I keep thinking of remote nonces as the verification nonces because when you receive that partial signature, that's the nonce that you receive alongside it to verify that partial signature – but that's the opposite of the intended nomenclature lol

File: bolt-simple-taproot.md
Comment (arik-so) [reply bolt-simple-taproot.md:417]: So like you did, I also wanna break it down into the two components involved: one is sending the remote nonce as part of the partial signature, and the other is sending the local nonce just in time.

I don't think we have a disagreement regarding the utility of including the remote nonce with the partial signature, so really the discussion is about whether the first local nonce should be included in the open_channel message, or in funding_created. From an implementation perspective, there is, as you said, not much of a difference, because either way, the sender needs to cache the nonce until they receive a partial signature to verify the correctness. But there is a difference in the semantics.

Specifically, I believe that the nonce/partial signature dance is essentially a challenge response mechanism. Alice signals to Bob that she's ready for a signature, and here's the nonce that signature should aggregate against. When should that nonce be sent? In my opinion, when readiness for that signature is signaled, i. e. in the message that immediately precedes the partial signature being sent by the remote party.

Separately, Wilmer made the point that the open_channel message is really about setting up the channel parameters, and the nonce for the first signature really isn't a parameter of the channel.

So these are the reasons that we think it makes semantically more sense for the local nonce to be moved to the message that we moved them to, but again, as you pointed out, there's no big difference from the standpoint of actually needing to implement it — just from understanding the rationale behind why the implementation needs to be a certain way.

File: bolt-simple-taproot.md
Comment (arik-so) [reply bolt-simple-taproot.md]: I am not entirely sure I agree with your last point. If following a disconnect I need to make the final sig, I have no choice but to either remember the last (remote) nonce you sent, or to at the very least remember the last aggregate nonce I calculated with the remote nonce you sent. So that requirement is there either way, no? It's either that or storing the full signature.

**Comment by arik-so** [comment]
btw, we can remove all the low-level recommendations and introduce verbiage that discriminates between calculation and cache retrieval of the combined funding key

File: bolt-simple-taproot.md
Comment (halseth) [inline bolt-simple-taproot.md:261]: 33-byte?

File: bolt-simple-taproot.md
Comment (halseth) [inline bolt-simple-taproot.md]: s/chanenls/channels

File: bolt-simple-taproot.md
Comment (halseth) [reply bolt-simple-taproot.md:395]: Agree on naming the nonce field according to what they actually are. Makes it much easier to just glance at the TLVs and know what the data is.

File: bolt-simple-taproot.md
Comment (halseth) [reply bolt-simple-taproot.md:417]: > So I think this is the main thing I'm contesting here: is sending nonces later in the process actually simpler? Or is it just shifting things around w/ minimal gain?

It is definitely just shifting things around, but I think it has some merit. Not sending nonces upfront makes more logical sense in my head, as why would you send them before you know whether you'll ever get so far in the funding process that you'll actually need them?

File: bolt-simple-taproot.md
Comment (arik-so) [reply bolt-simple-taproot.md:261]: No, the secret nonce components are 32 bytes each.

File: bolt-simple-taproot.md
Comment (arik-so) [reply bolt-simple-taproot.md]: Will update, thanks!

File: bolt-simple-taproot.md
Comment (arik-so) [reply bolt-simple-taproot.md:417]: Good point! 

By the way, if we all agree that remote nonces should be made part of the partial signature, I think it would be somewhat less confusing to external readers if we were referring to the nonce in singular, given that it's really only the local nonce that has its own dedicated TLV field.

File: bolt-simple-taproot.md
Comment (arik-so) [reply bolt-simple-taproot.md]: We ended up going with `next_local_nonce`.

File: bolt-simple-taproot.md
Comment (arik-so) [reply bolt-simple-taproot.md:395]: We ended up going with `next_local_nonce` for the standalone nonce's name.

File: bolt-simple-taproot.md
Comment (arik-so) [reply bolt-simple-taproot.md:679]: We split it back out from `channel_ready`.

File: bolt-simple-taproot.md
Comment (arik-so) [reply bolt-simple-taproot.md]: We split `channel_ready` and `channel_reestablish` back up again, and incorporated the commitment signature retransmission requirements in the latter section.

**Pull request merged**
